File: 1_finding_exit_from_maze/reachability.cpp
#include<bits/stdc++.h>

using std::vector;
using std::pair;

bool explore(vector<vector<int> > &adj,bool visited[],int x,int y)
{
    if(x==y)
    return true;
    visited[x]=true;
    for(auto it=adj[x].begin();it!=adj[x].end();it++)
    if(!visited[*it])
   { 
       if(explore(adj,visited,*it,y))
       return true;
   }
   return false;
}
int reach(vector<vector<int> > &adj, int x, int y) {
  bool visited[adj.size()]={false};
  if(explore(adj,visited,x,y)==true)
  return 1;
  else
  return 0;
}

int main() {
  size_t n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  int x, y;
  std::cin >> x >> y;
  std::cout << reach(adj, x - 1, y - 1);
}

File: week1_graph_decomposition1/1_finding_exit_from_maze/Reachability.java
import java.util.ArrayList;
import java.util.Scanner;

public class Reachability {
    public static boolean explore(ArrayList<Integer>[] adj,boolean[] visited,int x,int y)
    {
        if(x==y)
        return true;
        visited[x]=true;
        for(int i = 0; i < adj[x].size(); i++)
        if(!visited[adj[x].get(i)])
        if(explore(adj,visited,adj[x].get(i),y))
        return true;
        
        return false;
    }
    private static int reach(ArrayList<Integer>[] adj, int x, int y) {
        boolean[] visited=new boolean[adj.length];
        if(explore(adj,visited,x,y)==true)
        return 1;
        else
        return 0;
    }


    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        int x = scanner.nextInt() - 1;
        int y = scanner.nextInt() - 1;
        System.out.println(reach(adj, x, y));
    }
}
